src/Validation.ts

```diff
@@ -0,0 +1,215 @@
+type Fork = (
+    reject: (value: Error) => void,
+    resolve: (value: Valid | Invalid) => void
+) => void;
+
+export interface InvalidResult {
+    value: any;
+    message: string;
+    key?: string[];
+}
+
+export interface Valid {
+    isValid: true;
+    x: undefined;
+    fork: undefined;
+    and(
+        other: Valid | Invalid | AsyncValidation
+    ): Valid | Invalid | AsyncValidation;
+    and(other: Valid): Valid;
+    and(other: Invalid): Invalid;
+    and(other: AsyncValidation): AsyncValidation;
+    or(other: Valid | Invalid | AsyncValidation): Valid | AsyncValidation;
+    or(other: Valid | Invalid): Valid;
+    or(other: AsyncValidation): AsyncValidation;
+    format: (
+        fn: (
+            invalidResult: InvalidResult,
+            index: number,
+            all: InvalidResult[]
+        ) => InvalidResult
+    ) => Valid;
+    getResult: () => void;
+}
+
+export interface Invalid {
+    x: InvalidResult[];
+    isValid: false;
+    fork: undefined;
+    and(other: Valid | Invalid | AsyncValidation): Invalid | AsyncValidation;
+    and(other: AsyncValidation): AsyncValidation;
+    and(other: Valid | Invalid): Invalid;
+
+    or(
+        other: Valid | Invalid | AsyncValidation
+    ): Valid | Invalid | AsyncValidation;
+    or(other: AsyncValidation): AsyncValidation;
+    or(other: Valid): Valid;
+    or(other: Invalid): Invalid;
+    format: (
+        fn: (
+            invalidResult: InvalidResult,
+            index: number,
+            all: InvalidResult[]
+        ) => InvalidResult
+    ) => Invalid;
+    getResult: () => InvalidResult[];
+}
+
+export type SyncValidation = Valid | Invalid;
+
+export interface AsyncValidation {
+    x: undefined;
+    fork: Fork;
+    isValid: undefined;
+    and(v: Valid | Invalid | AsyncValidation): AsyncValidation;
+    or(v: Valid | Invalid | AsyncValidation): AsyncValidation;
+    format: (
+        fn: (
+            invalidResult: InvalidResult,
+            index: number,
+            all: InvalidResult[]
+        ) => InvalidResult
+    ) => AsyncValidation;
+    getResult: () => Promise<void | InvalidResult[]>;
+}
+
+const isValid = (value: Valid | Invalid | AsyncValidation): value is Valid =>
+    value.isValid;
+
+const isInvalid = (
+    value: Valid | Invalid | AsyncValidation
+): value is Invalid => value.isValid === false;
+
+const isAsync = (
+    value: Valid | Invalid | AsyncValidation
+): value is AsyncValidation => !!value.fork;
+
+// Valid will hold no value
+export const Valid = (): Valid => ({
```

Comment:
having both a function and a type with the same name is confusing IMHO

Reply:
Since the function is a factory to generate the type, I do not see an issue here.